controller/src/util/instance_action.rs

```diff
@@ -472,17 +474,41 @@ pub async fn handle_instance_change(
         .await?
     }
 
-    let nodes_to_add = nodes_to_act_on
+    // Check how many nodes already have pods
+    let num_nodes_already_with_pods: i32 = nodes_to_act_on
         .iter()
-        .filter_map(|(node, context)| {
-            if ((context.action) == PodAction::Add) | ((context.action) == PodAction::RemoveAndAdd)
-            {
-                Some(node.to_string())
+        .filter_map(|(_, context)| {
+            if (context.action) == PodAction::NoAction {
+                Some(1)
             } else {
                 None
             }
         })
-        .collect::<Vec<String>>();
+        .sum();
+
+    // Assert that more nodes don't receive pods than specified by capacity
+    let num_nodes_to_receive_pod = capacity as i32 - num_nodes_already_with_pods;
+    trace!("handle_instance_change - capacity [{}] - number of nodes already with pods [{}] = [{}] node[s] to receive pods", capacity, num_nodes_already_with_pods, num_nodes_to_receive_pod);
+    let nodes_to_add: Vec<String> = if num_nodes_to_receive_pod < 0 {
+        error!(
```

Comment:
Would this not happen on capacity change? Or would that trigger different handler and it would not come here?

Reply:
The only way to change capacity currently is to modify the configuration. Currently, when the agent sees a configuration modification, it deletes the associated instance[s] and all device plugins and recreates it. All pods will be brought down by the controller (once the instance is deleted) and this code will be run through again for an Instance added event.